src/OpenLoco/src/World/CompanyAi.cpp

```diff
@@ -658,36 +674,500 @@ namespace OpenLoco
+
+            for (const auto dockObjId : docks)
+            {
+                if (dockObjId == 0xFFU)
+                {
+                    break;
+                }
+                auto* dockObj = ObjectManager::get<DockObject>(dockObjId);
+
+                if (bestDesignYear < dockObj->designedYear)
+                {
+                    bestDesignYear = dockObj->designedYear;
+                    chosenStationObject = dockObjId;
+                }
+            }
+            if (bestDesignYear == -1)
+            {
+                return true;
+            }
+            thought.stationObjId = chosenStationObject;
+            thought.signalObjId = 0xFFU;
+            return false;
+        }
+        else if (thought.trackObjId & (1U << 7))
+        {
+            std::array<uint8_t, 17> roadStations{};
+            Ui::Windows::Construction::Common::refreshStationList(roadStations.data(), thought.trackObjId & ~(1U << 7), TransportMode::road);
+            int16_t bestDesignYear = -1;
+            bool unk = false;
+            for (const auto roadStationObjId : roadStations)
+            {
+                if (roadStationObjId == 0xFFU)
+                {
+                    break;
+                }
+                auto* roadStationObj = ObjectManager::get<RoadStationObject>(roadStationObjId);
+                if (roadStationObj->hasFlags(RoadStationFlags::passenger)
+                    && roadStationObj->cargoType != thought.cargoType)
+                {
+                    continue;
+                }
+                if (roadStationObj->hasFlags(RoadStationFlags::freight)
+                    && roadStationObj->cargoType == thought.cargoType) // Why??
+                {
+                    continue;
+                }
+
+                bool alwaysSelect = false;
+                if (thoughtTypeHasFlags(thought.type, ThoughtTypeFlags::unk8))
+                {
+                    if (roadStationObj->hasFlags(RoadStationFlags::roadEnd))
+                    {
+                        if (!unk)
+                        {
+                            alwaysSelect = true;
+                        }
+                        unk = true;
+                    }
+                    else
+                    {
+                        if (unk)
+                        {
+                            continue;
+                        }
+                    }
+                }
+                else
+                {
+                    if (roadStationObj->hasFlags(RoadStationFlags::roadEnd))
+                    {
+                        if (unk)
+                        {
+                            continue;
+                        }
+                    }
+                    else
+                    {
+                        if (!unk)
+                        {
+                            alwaysSelect = true;
+                        }
+                        unk = true;
+                    }
+                }
```

Comment:
Can we deduplicate this? One branch seems to be the inverse of the other.

Reply:
would it make more sense if it was the following:
```
bool hasRequiredRoadEnd = roadStationObj->hasFlags(RoadStationFlags::roadEnd) == thoughtTypeHasFlags(thought.type, ThoughtTypeFlags::unk8);

// If we have previously used a fallback we want to remove the fallback and use
// the ideal selection
bool alwaysSelect = hasRequiredRoadEnd && !hadIdealSelection;

// We have now entered ideal selection mode so can remove non-ideal stations
// from potential selection
if (hadIdealSelection && !hasRequiredRoadEnd){
    continue;
}
hadIdealSelection |= hasRequiredRoadEnd;
```
I'm pretty sure this is the same logic.